File: typescript/agentkit/src/action-providers/pendle/pendleActionProvider.ts
import { z } from "zod";
import { encodeFunctionData, Hex, parseUnits } from "viem";
import { erc20Abi } from "viem";
import { ActionProvider } from "../actionProvider";
import { EvmWalletProvider } from "../../wallet-providers";
import { CreateAction } from "../actionDecorator";
import { approve } from "../../utils";
import { Network } from "../../network";
import {
  PENDLE_ROUTER_V4,
  PENDLE_API_BASE,
  PENDLE_ROUTER_ABI,
  SUPPORTED_CHAIN_IDS,
} from "./constants";
import {
  BuyPtSchema,
  SellPtSchema,
  BuyYtSchema,
  SellYtSchema,
  AddLiquiditySchema,
  RemoveLiquiditySchema,
  ClaimRewardsSchema,
  ListMarketsSchema,
} from "./schemas";

const NATIVE_TOKEN = "0x0000000000000000000000000000000000000000";

const PENDLE_SUPPORTED_NETWORKS = ["base-mainnet"];

/**
 * PendleActionProvider enables interaction with Pendle Finance on Base.
 *
 * Pendle splits yield-bearing tokens into PT (Principal Token) and YT (Yield Token),
 * enabling fixed-yield and yield-speculation strategies. This provider uses the
 * Pendle Hosted SDK API for optimal execution (better gas, aggregator routing,
 * limit order matching).
 */
export class PendleActionProvider extends ActionProvider<EvmWalletProvider> {
  constructor() {
    super("pendle", []);
  }

  /**
   * Calls the Pendle Convert API to get transaction calldata.
   */
  private async callConvertApi(
    chainId: number,
    receiver: string,
    slippage: number,
    inputs: Array<{ tokenAddress: string; amount: string }>,
    outputs: Array<{ tokenAddress: string }>,
  ): Promise<{
    tx: { to: string; data: string; value: string };
    requiredApprovals: Array<{ tokenAddress: string; spender: string; amount: string }>;
  }> {
    const url = `${PENDLE_API_BASE}/v3/sdk/${chainId}/convert`;
    const body = {
      receiver,
      slippage,
      enableAggregator: true,
      inputs,
      outputs,
    };

    const response = await fetch(url, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    });

    if (!response.ok) {
      const text = await response.text();
      throw new Error(`Pendle API error (${response.status}): ${text}`);
    }

    return response.json();
  }

  /**
   * Resolves token decimals — returns 18 for native ETH.
   */
  private async getDecimals(wallet: EvmWalletProvider, tokenAddress: string): Promise<number> {
    if (tokenAddress.toLowerCase() === NATIVE_TOKEN) return 18;
    return (await wallet.readContract({
      address: tokenAddress as Hex,
      abi: erc20Abi,
      functionName: "decimals",
      args: [],
    })) as number;
  }

  /**
   * Gets the chain ID for the current network.
   */
  private getChainId(wallet: EvmWalletProvider): number {
    const network = wallet.getNetwork();
    const chainId = SUPPORTED_CHAIN_IDS[network.networkId!];
    if (!chainId) throw new Error(`Unsupported network: ${network.networkId}`);
    return chainId;
  }

  /**
   * Fetches market info from Pendle API to resolve PT/YT/LP addresses.
   */
  private async getMarketInfo(
    chainId: number,
    marketAddress: string,
  ): Promise<{
    pt: string;
    yt: string;
    sy: string;
    underlyingAsset: string;
    expiry: string;
    name: string;
  }> {
    const url = `${PENDLE_API_BASE}/v1/sdk/${chainId}/markets/${marketAddress}`;
    const response = await fetch(url);
    if (!response.ok) {
      throw new Error(`Failed to fetch market info: ${response.status}`);
    }
    return response.json();
  }

  /**
   * Executes a convert operation via the Pendle API and submits the transaction.
   */
  private async executeConvert(
    wallet: EvmWalletProvider,
    slippage: number,
    inputs: Array<{ tokenAddress: string; amount: string }>,
    outputs: Array<{ tokenAddress: string }>,
  ): Promise<{ txHash: string; receipt: unknown }> {
    const chainId = this.getChainId(wallet);
    const receiver = await wallet.getAddress();

    const result = await this.callConvertApi(chainId, receiver, slippage, inputs, outputs);

    // Handle required approvals
    for (const approval of result.requiredApprovals || []) {
      const approvalResult = await approve(
        wallet,
        approval.tokenAddress,
        approval.spender,
        BigInt(approval.amount),
      );
      if (approvalResult.startsWith("Error")) {
        throw new Error(`Token approval failed: ${approvalResult}`);
      }
    }

    // Submit the transaction
    const txHash = await wallet.sendTransaction({
      to: result.tx.to as `0x${string}`,
      data: result.tx.data as `0x${string}`,
      value: BigInt(result.tx.value || "0"),
    });

    const receipt = await wallet.waitForTransactionReceipt(txHash);
    return { txHash, receipt };
  }

  // ─── Actions ────────────────────────────────────────────────

  @CreateAction({
    name: "buy_pt",
    description: `
Buy PT (Principal Token) on Pendle Finance to lock in a fixed yield.

PT represents the principal portion of a yield-bearing token. It trades at a discount
before maturity and is redeemable 1:1 for the underlying asset at maturity.
The discount = your fixed yield.

It takes:
- market: The Pendle market address (find markets using the list_markets action)
- tokenIn: The token to swap from (e.g., WETH, USDC address). Use 0x0000000000000000000000000000000000000000 for native ETH
- amount: Amount of input tokens in whole units (e.g., '0.1' for 0.1 WETH)
- slippage: Slippage tolerance (default 0.01 = 1%)

Example: Buy PT with 0.1 ETH on the yoETH market to lock in ~5% fixed yield.
`,
    schema: BuyPtSchema,
  })
  async buyPt(wallet: EvmWalletProvider, args: z.infer<typeof BuyPtSchema>): Promise<string> {
    try {
      const chainId = this.getChainId(wallet);
      const decimals = await this.getDecimals(wallet, args.tokenIn);
      const atomicAmount = parseUnits(args.amount, decimals).toString();
      const marketInfo = await this.getMarketInfo(chainId, args.market);

      const { txHash, receipt } = await this.executeConvert(
        wallet,
        args.slippage,
        [{ tokenAddress: args.tokenIn, amount: atomicAmount }],
        [{ tokenAddress: marketInfo.pt }],
      );

      return `Successfully bought PT on Pendle market ${args.market}.\nInput: ${args.amount} tokens\nTransaction: ${txHash}\nReceipt: ${JSON.stringify(receipt)}`;
    } catch (error) {
      return `Error buying PT on Pendle: ${error}`;
    }
  }

  @CreateAction({
    name: "sell_pt",
    description: `
Sell PT (Principal Token) on Pendle Finance back to an underlying token.

Use this to exit a fixed-yield position before maturity, or to redeem after maturity.

It takes:
- market: The Pendle market address
- tokenOut: The token to receive (e.g., WETH, USDC address). Use 0x0000000000000000000000000000000000000000 for native ETH
- amount: Amount of PT to sell, in whole units
- slippage: Slippage tolerance (default 0.01 = 1%)
`,
    schema: SellPtSchema,
  })
  async sellPt(wallet: EvmWalletProvider, args: z.infer<typeof SellPtSchema>): Promise<string> {
    try {
      const chainId = this.getChainId(wallet);
      const marketInfo = await this.getMarketInfo(chainId, args.market);

      // PT tokens have 18 decimals
      const atomicAmount = parseUnits(args.amount, 18).toString();

      const { txHash, receipt } = await this.executeConvert(
        wallet,
        args.slippage,
        [{ tokenAddress: marketInfo.pt, amount: atomicAmount }],
        [{ tokenAddress: args.tokenOut }],
      );

      return `Successfully sold PT on Pendle market ${args.market}.\nSold: ${args.amount} PT\nTransaction: ${txHash}\nReceipt: ${JSON.stringify(receipt)}`;
    } catch (error) {
      return `Error selling PT on Pendle: ${error}`;
    }
  }

  @CreateAction({
    name: "buy_yt",
    description: `
Buy YT (Yield Token) on Pendle Finance to speculate on yield going up.

YT holders receive all yield generated by the underlying asset until maturity.
If you expect yields to increase, buying YT lets you profit from that.

It takes:
- market: The Pendle market address (find markets using the list_markets action)
- tokenIn: The token to swap from. Use 0x0000000000000000000000000000000000000000 for native ETH
- amount: Amount of input tokens in whole units
- slippage: Slippage tolerance (default 0.01 = 1%)
`,
    schema: BuyYtSchema,
  })
  async buyYt(wallet: EvmWalletProvider, args: z.infer<typeof BuyYtSchema>): Promise<string> {
    try {
      const chainId = this.getChainId(wallet);
      const decimals = await this.getDecimals(wallet, args.tokenIn);
      const atomicAmount = parseUnits(args.amount, decimals).toString();
      const marketInfo = await this.getMarketInfo(chainId, args.market);

      const { txHash, receipt } = await this.executeConvert(
        wallet,
        args.slippage,
        [{ tokenAddress: args.tokenIn, amount: atomicAmount }],
        [{ tokenAddress: marketInfo.yt }],
      );

      return `Successfully bought YT on Pendle market ${args.market}.\nInput: ${args.amount} tokens\nTransaction: ${txHash}\nReceipt: ${JSON.stringify(receipt)}`;
    } catch (error) {
      return `Error buying YT on Pendle: ${error}`;
    }
  }

  @CreateAction({
    name: "sell_yt",
    description: `
Sell YT (Yield Token) on Pendle Finance back to an underlying token.

Use this to exit a yield speculation position.

It takes:
- market: The Pendle market address
- tokenOut: The token to receive. Use 0x0000000000000000000000000000000000000000 for native ETH
- amount: Amount of YT to sell, in whole units
- slippage: Slippage tolerance (default 0.01 = 1%)
`,
    schema: SellYtSchema,
  })
  async sellYt(wallet: EvmWalletProvider, args: z.infer<typeof SellYtSchema>): Promise<string> {
    try {
      const chainId = this.getChainId(wallet);
      const marketInfo = await this.getMarketInfo(chainId, args.market);
      const atomicAmount = parseUnits(args.amount, 18).toString();

      const { txHash, receipt } = await this.executeConvert(
        wallet,
        args.slippage,
        [{ tokenAddress: marketInfo.yt, amount: atomicAmount }],
        [{ tokenAddress: args.tokenOut }],
      );

      return `Successfully sold YT on Pendle market ${args.market}.\nSold: ${args.amount} YT\nTransaction: ${txHash}\nReceipt: ${JSON.stringify(receipt)}`;
    } catch (error) {
      return `Error selling YT on Pendle: ${error}`;
    }
  }

  @CreateAction({
    name: "add_liquidity",
    description: `
Add liquidity to a Pendle market to earn swap fees and PENDLE rewards.

Pendle LP positions earn from three sources: swap fees, underlying yield, and PENDLE incentives.

It takes:
- market: The Pendle market address
- tokenIn: The token to provide as liquidity. Use 0x0000000000000000000000000000000000000000 for native ETH
- amount: Amount of tokens to add, in whole units
- slippage: Slippage tolerance (default 0.01 = 1%)
`,
    schema: AddLiquiditySchema,
  })
  async addLiquidity(
    wallet: EvmWalletProvider,
    args: z.infer<typeof AddLiquiditySchema>,
  ): Promise<string> {
    try {
      const chainId = this.getChainId(wallet);
      const decimals = await this.getDecimals(wallet, args.tokenIn);
      const atomicAmount = parseUnits(args.amount, decimals).toString();

      // LP token address is the market address itself
      const { txHash, receipt } = await this.executeConvert(
        wallet,
        args.slippage,
        [{ tokenAddress: args.tokenIn, amount: atomicAmount }],
        [{ tokenAddress: args.market }],
      );

      return `Successfully added liquidity to Pendle market ${args.market}.\nInput: ${args.amount} tokens\nTransaction: ${txHash}\nReceipt: ${JSON.stringify(receipt)}`;
    } catch (error) {
      return `Error adding liquidity on Pendle: ${error}`;
    }
  }

  @CreateAction({
    name: "remove_liquidity",
    description: `
Remove liquidity from a Pendle market back to a single token.

It takes:
- market: The Pendle market address
- tokenOut: The token to receive back. Use 0x0000000000000000000000000000000000000000 for native ETH
- amount: Amount of LP tokens to remove, in whole units
- slippage: Slippage tolerance (default 0.01 = 1%)
`,
    schema: RemoveLiquiditySchema,
  })
  async removeLiquidity(
    wallet: EvmWalletProvider,
    args: z.infer<typeof RemoveLiquiditySchema>,
  ): Promise<string> {
    try {
      const atomicAmount = parseUnits(args.amount, 18).toString();

      const { txHash, receipt } = await this.executeConvert(
        wallet,
        args.slippage,
        [{ tokenAddress: args.market, amount: atomicAmount }],
        [{ tokenAddress: args.tokenOut }],
      );

      return `Successfully removed liquidity from Pendle market ${args.market}.\nRemoved: ${args.amount} LP\nTransaction: ${txHash}\nReceipt: ${JSON.stringify(receipt)}`;
    } catch (error) {
      return `Error removing liquidity on Pendle: ${error}`;
    }
  }

  @CreateAction({
    name: "claim_rewards",
    description: `
Claim accrued interest, yield, and PENDLE rewards from Pendle positions.

Pendle positions accrue rewards over time:
- YT holders earn yield from the underlying asset
- LP providers earn swap fees and PENDLE incentives
- SY holders may earn interest

Provide the addresses of the positions you want to claim from.

It takes:
- syAddresses: Array of SY token addresses (optional)
- ytAddresses: Array of YT token addresses (optional)
- marketAddresses: Array of market/LP addresses (optional)
`,
    schema: ClaimRewardsSchema,
  })
  async claimRewards(
    wallet: EvmWalletProvider,
    args: z.infer<typeof ClaimRewardsSchema>,
  ): Promise<string> {
    try {
      const userAddress = await wallet.getAddress();

      const data = encodeFunctionData({
        abi: PENDLE_ROUTER_ABI,
        functionName: "redeemDueInterestAndRewards",
        args: [
          userAddress as `0x${string}`,
          (args.syAddresses || []) as `0x${string}`[],
          (args.ytAddresses || []) as `0x${string}`[],
          (args.marketAddresses || []) as `0x${string}`[],
        ],
      });

      const txHash = await wallet.sendTransaction({
        to: PENDLE_ROUTER_V4 as `0x${string}`,
        data,
      });

      const receipt = await wallet.waitForTransactionReceipt(txHash);

      return `Successfully claimed Pendle rewards.\nTransaction: ${txHash}\nReceipt: ${JSON.stringify(receipt)}`;
    } catch (error) {
      return `Error claiming Pendle rewards: ${error}`;
    }
  }

  @CreateAction({
    name: "list_markets",
    description: `
List active Pendle markets on Base with their current APYs, TVL, and maturity dates.

Returns a formatted list of markets including:
- Market name and address
- Underlying asset
- Fixed APY (implied from PT price)
- TVL
- Maturity date

No inputs required.
`,
    schema: ListMarketsSchema,
  })
  async listMarkets(wallet: EvmWalletProvider): Promise<string> {
    try {
      const chainId = this.getChainId(wallet);
      const url = `${PENDLE_API_BASE}/v2/markets/all?chainId=${chainId}`;
      const response = await fetch(url);

      if (!response.ok) {
        throw new Error(`Pendle API error: ${response.status}`);
      }

      const data = await response.json();
      const markets = (data.results || data) as Array<{
        address: string;
        name: string;
        expiry: string;
        pt: { address: string };
        yt: { address: string };
        underlyingAsset: { symbol: string; address: string };
        aggregatedApy: number;
        impliedApy: number;
        tvl: { usd: number };
        liquidity: { usd: number };
      }>;

      if (!markets || markets.length === 0) {
        return "No active Pendle markets found on Base.";
      }

      const lines = markets
        .filter((m) => new Date(m.expiry) > new Date())
        .sort((a, b) => (b.tvl?.usd || 0) - (a.tvl?.usd || 0))
        .slice(0, 15)
        .map((m) => {
          const expiry = new Date(m.expiry).toLocaleDateString();
          const tvl = m.tvl?.usd ? `$${(m.tvl.usd / 1e6).toFixed(2)}M` : "N/A";
          const fixedApy = m.impliedApy ? `${(m.impliedApy * 100).toFixed(2)}%` : "N/A";
          return `- ${m.name} | Market: ${m.address} | Fixed APY: ${fixedApy} | TVL: ${tvl} | Expires: ${expiry}`;
        });

      return `Active Pendle Markets on Base:\n\n${lines.join("\n")}`;
    } catch (error) {
      return `Error listing Pendle markets: ${error}`;
    }
  }

  supportsNetwork = (network: Network) =>
    network.protocolFamily === "evm" && PENDLE_SUPPORTED_NETWORKS.includes(network.networkId!);
}

export const pendleActionProvider = () => new PendleActionProvider();

File: python/coinbase-agentkit/coinbase_agentkit/action_providers/pendle/pendle_action_provider.py
"""Pendle V2 action provider for trading Principal Tokens (PT) via the hosted SDK."""

from typing import Any

from web3 import Web3

from ...network import Network
from ...wallet_providers import EvmWalletProvider
from ..action_decorator import create_action
from ..action_provider import ActionProvider
from .constants import PENDLE_ROUTER_ADDRESS, SUPPORTED_NETWORKS
from .schemas import (
    PendleMarketInfoSchema,
    PendleSwapPtForTokenSchema,
    PendleSwapTokenForPtSchema,
)
from .utils import (
    approve_token,
    format_amount_from_decimals,
    format_amount_with_decimals,
    get_chain_id_for_network,
    get_token_balance,
    get_token_decimals,
    get_token_symbol,
    pendle_convert,
    pendle_get_market,
)


class PendleActionProvider(ActionProvider[EvmWalletProvider]):
    """Provides actions for trading Pendle V2 PT positions via the hosted SDK."""

    def __init__(self):
        """Initialize the Pendle action provider."""
        super().__init__("pendle", [])

    def supports_network(self, network: Network) -> bool:
        """Check if the current network is supported by Pendle V2.

        Args:
            network: The network to check.

        Returns:
            bool: True if the network is a supported Pendle origin chain.

        """
        return network.protocol_family == "evm" and network.network_id in SUPPORTED_NETWORKS

    def _submit_pendle_route(
        self,
        wallet_provider: EvmWalletProvider,
        sdk_response: dict[str, Any],
    ) -> str:
        """Approve required tokens then submit the route's prepared transaction.

        Args:
            wallet_provider: The wallet to send from.
            sdk_response: The full response object from Pendle's /convert endpoint.

        Returns:
            str: The submitted transaction hash.

        """
        for approval in sdk_response.get("requiredApprovals", []):
            approve_token(
                wallet_provider,
                approval["token"],
                PENDLE_ROUTER_ADDRESS,
                int(approval["amount"]),
            )

        route = sdk_response["routes"][0]
        tx = route["tx"]
        params = {
            "to": Web3.to_checksum_address(tx["to"]),
            "data": tx["data"],
            "value": int(tx.get("value", 0)),
        }

        tx_hash = wallet_provider.send_transaction(params)
        wallet_provider.wait_for_transaction_receipt(tx_hash)
        return tx_hash

    @create_action(
        name="swap_exact_token_for_pt",
        description="""
This tool buys Pendle Principal Tokens (PT) on a Pendle V2 market by spending an
underlying ERC-20 (e.g. USDC, USDT, WETH).

It takes:
- market_address: Address of the Pendle market on the current chain
- token_in_address: Address of the input ERC-20 to spend
- amount: Human-readable amount of token_in to spend (e.g. `100` for 100 USDC)
- slippage: Slippage tolerance as a decimal, default 0.005 (0.5%)

Important notes:
- The wallet must hold enough of token_in to cover the amount.
- PT trades at a discount to the underlying; you receive PT atomic units, not the underlying.
- The Pendle Router is approved to spend token_in before the swap.
- Calldata is fetched from Pendle's hosted SDK at runtime to handle the V2 router's complex routing and aggregator selection.
""",
        schema=PendleSwapTokenForPtSchema,
    )
    def swap_exact_token_for_pt(
        self, wallet_provider: EvmWalletProvider, args: dict[str, Any]
    ) -> str:
        """Buy Pendle PT with an ERC-20 token.

        Args:
            wallet_provider: The wallet to swap from.
            args: The swap parameters.

        Returns:
            str: A message describing the result of the swap.

        """
        try:
            validated_args = PendleSwapTokenForPtSchema(**args)
            network = wallet_provider.get_network()

            if not self.supports_network(network):
                return f"Error: Network {network.network_id} is not supported by Pendle"

            try:
                chain_id = get_chain_id_for_network(network)
            except ValueError as e:
                return f"Error: {e!s}"

            try:
                market = pendle_get_market(chain_id, validated_args.market_address)
            except Exception as e:
                return f"Error fetching Pendle market {validated_args.market_address}: {e!s}"

            pt_address = market.get("pt", {}).get("address")
            if not pt_address:
                return f"Error: Pendle market {validated_args.market_address} did not return a PT address"

            token_in = Web3.to_checksum_address(validated_args.token_in_address)

            try:
                decimals = get_token_decimals(wallet_provider, token_in)
                amount_atomic = format_amount_with_decimals(validated_args.amount, decimals)
            except Exception as e:
                return f"Error reading token info for {validated_args.token_in_address}: {e!s}"

            try:
                wallet_balance = get_token_balance(wallet_provider, token_in)
                if wallet_balance < amount_atomic:
                    human_balance = format_amount_from_decimals(wallet_balance, decimals)
                    return (
                        f"Error: Insufficient balance. Wallet has {human_balance} of token "
                        f"{validated_args.token_in_address}, but trying to spend {validated_args.amount}"
                    )
            except Exception as e:
                return f"Error checking token balance: {e!s}"

            try:
                sdk_response = pendle_convert(
                    chain_id=chain_id,
                    receiver=wallet_provider.get_address(),
                    slippage=validated_args.slippage,
                    token_in_address=token_in,
                    amount_in_atomic=amount_atomic,
                    token_out_address=pt_address,
                )
            except Exception as e:
                return f"Error fetching swap calldata from Pendle SDK: {e!s}"

            try:
                tx_hash = self._submit_pendle_route(wallet_provider, sdk_response)
            except Exception as e:
                return f"Error executing Pendle swap: {e!s}"

            try:
                token_in_symbol = get_token_symbol(wallet_provider, token_in)
            except Exception:
                token_in_symbol = "tokens"

            market_symbol = market.get("symbol", validated_args.market_address)
            return (
                f"Successfully swapped {validated_args.amount} {token_in_symbol} for PT in "
                f"Pendle market {market_symbol}.\n"
                f"Transaction hash: {tx_hash}"
            )
        except Exception as e:
            return f"Error swapping for PT on Pendle: {e!s}"

    @create_action(
        name="swap_exact_pt_for_token",
        description="""
This tool sells Pendle Principal Tokens (PT) on a Pendle V2 market for an
ERC-20 (e.g. USDC, USDT, WETH).

It takes:
- market_address: Address of the Pendle market on the current chain
- token_out_address: Address of the ERC-20 to receive
- pt_amount: Human-readable amount of PT to sell (e.g. `100`)
- slippage: Slippage tolerance as a decimal, default 0.005 (0.5%)

Important notes:
- The wallet must hold enough PT to cover pt_amount.
- Selling PT before maturity may incur a market discount; selling at/after maturity returns the underlying 1:1 (less fees).
- The Pendle Router is approved to spend PT before the swap.
- Calldata is fetched from Pendle's hosted SDK at runtime.
""",
        schema=PendleSwapPtForTokenSchema,
    )
    def swap_exact_pt_for_token(
        self, wallet_provider: EvmWalletProvider, args: dict[str, Any]
    ) -> str:
        """Sell Pendle PT for an ERC-20 token.

        Args:
            wallet_provider: The wallet to swap from.
            args: The swap parameters.

        Returns:
            str: A message describing the result of the swap.

        """
        try:
            validated_args = PendleSwapPtForTokenSchema(**args)
            network = wallet_provider.get_network()

            if not self.supports_network(network):
                return f"Error: Network {network.network_id} is not supported by Pendle"

            try:
                chain_id = get_chain_id_for_network(network)
            except ValueError as e:
                return f"Error: {e!s}"

            try:
                market = pendle_get_market(chain_id, validated_args.market_address)
            except Exception as e:
                return f"Error fetching Pendle market {validated_args.market_address}: {e!s}"

            pt_address = market.get("pt", {}).get("address")
            if not pt_address:
                return f"Error: Pendle market {validated_args.market_address} did not return a PT address"

            pt_address = Web3.to_checksum_address(pt_address)
            token_out = Web3.to_checksum_address(validated_args.token_out_address)

            try:
                pt_decimals = get_token_decimals(wallet_provider, pt_address)
                pt_atomic = format_amount_with_decimals(validated_args.pt_amount, pt_decimals)
            except Exception as e:
                return f"Error reading PT decimals: {e!s}"

            try:
                pt_balance = get_token_balance(wallet_provider, pt_address)
                if pt_balance < pt_atomic:
                    human_balance = format_amount_from_decimals(pt_balance, pt_decimals)
                    return (
                        f"Error: Insufficient PT balance. Wallet holds {human_balance} PT, "
                        f"but trying to sell {validated_args.pt_amount}"
                    )
            except Exception as e:
                return f"Error checking PT balance: {e!s}"

            try:
                sdk_response = pendle_convert(
                    chain_id=chain_id,
                    receiver=wallet_provider.get_address(),
                    slippage=validated_args.slippage,
                    token_in_address=pt_address,
                    amount_in_atomic=pt_atomic,
                    token_out_address=token_out,
                )
            except Exception as e:
                return f"Error fetching swap calldata from Pendle SDK: {e!s}"

            try:
                tx_hash = self._submit_pendle_route(wallet_provider, sdk_response)
            except Exception as e:
                return f"Error executing Pendle swap: {e!s}"

            try:
                token_out_symbol = get_token_symbol(wallet_provider, token_out)
            except Exception:
                token_out_symbol = "tokens"

            market_symbol = market.get("symbol", validated_args.market_address)
            return (
                f"Successfully sold {validated_args.pt_amount} PT for {token_out_symbol} "
                f"on Pendle market {market_symbol}.\n"
                f"Transaction hash: {tx_hash}"
            )
        except Exception as e:
            return f"Error selling PT on Pendle: {e!s}"

    @create_action(
        name="get_pendle_market_info",
        description="""
This tool reads metadata for a Pendle V2 market: PT/YT/SY addresses, expiry,
underlying asset, and symbol.

It takes:
- market_address: Address of the Pendle market on the current chain
""",
        schema=PendleMarketInfoSchema,
    )
    def get_pendle_market_info(
        self, wallet_provider: EvmWalletProvider, args: dict[str, Any]
    ) -> str:
        """Read details of a Pendle market.

        Args:
            wallet_provider: The wallet to read the network from.
            args: The market query parameters.

        Returns:
            str: A markdown summary of the market.

        """
        try:
            validated_args = PendleMarketInfoSchema(**args)
            network = wallet_provider.get_network()

            if not self.supports_network(network):
                return f"Error: Network {network.network_id} is not supported by Pendle"

            try:
                chain_id = get_chain_id_for_network(network)
            except ValueError as e:
                return f"Error: {e!s}"

            try:
                market = pendle_get_market(chain_id, validated_args.market_address)
            except Exception as e:
                return f"Error fetching Pendle market {validated_args.market_address}: {e!s}"

            pt = market.get("pt", {}) or {}
            yt = market.get("yt", {}) or {}
            sy = market.get("sy", {}) or {}
            underlying = market.get("underlyingAsset", {}) or {}

            return (
                f"# Pendle market {market.get('symbol', validated_args.market_address)}\n\n"
                f"- Address: {market.get('address')}\n"
                f"- Expiry: {market.get('expiry')}\n"
                f"- PT: {pt.get('address')} ({pt.get('symbol')})\n"
                f"- YT: {yt.get('address')} ({yt.get('symbol')})\n"
                f"- SY: {sy.get('address')} ({sy.get('symbol')})\n"
                f"- Underlying: {underlying.get('address')} ({underlying.get('symbol')})\n"
            )
        except Exception as e:
            return f"Error reading Pendle market info: {e!s}"


def pendle_action_provider() -> PendleActionProvider:
    """Create a new PendleActionProvider instance.

    Returns:
        PendleActionProvider: A new instance of the Pendle action provider.

    """
    return PendleActionProvider()
